src/cocoa/toga_cocoa/widgets/table.py

```diff
@@ -148,13 +170,38 @@ def change_source(self, source):
         self.table.reloadData()
 
     def insert(self, index, item):
-        self.table.reloadData()
+
+        # set parent = None if inserting to the root item
+        index_set = NSIndexSet.indexSetWithIndex(index)
+
+        self.table.insertRowsAtIndexes(
+            index_set,
+            withAnimation=NSTableViewAnimation.SlideDown
+        )
 
     def change(self, item):
-        self.table.reloadData()
+        row_index = self.interface.data.index(item)
+        row_indexes = NSIndexSet.indexSetWithIndex(row_index)
+        column_indexes = NSIndexSet.indexSetWithIndexesInRange(NSRange(0, len(self.columns)))
+        self.table.reloadDataForRowIndexes(
+            row_indexes,
+            columnIndexes=column_indexes
+        )
 
     def remove(self, item):
-        self.table.reloadData()
+        try:
+            # We can't get the index from self.interface.data because the
+            # row has already been removed. Instead we look up the index
+            # from an associated view.
+            index = self.table.rowForView(self._view_for_row[item])
```

Comment:
Should this be a `pop()`, rather a simple lookup? AFAICT, a removed item is never removed from `_view_for_row`.

Reply:
Yes, good catch. We don't want the dict to potentially grow indefinitely. I'll also clear `_view_for_row` when the entire table is cleared.